projectq/backends/_azure/_azure_quantum.py

```diff
@@ -0,0 +1,403 @@
+        qubits = len(qubit_mapping.keys())
+
+        if self._provider_id == IONQ_PROVIDER_ID:
+            return {
+                "qubits": qubits,
+                "circuit": self._circuit
+            }
+        elif self._provider_id == QUANTINUUM_PROVIDER_ID:
+            measurement_gates = ""
+
+            for measured_id in self._measured_ids:
+                qb_loc = self.main_engine.mapper.current_mapping[measured_id]
+                measurement_gates += "measure q[{0}] -> c[{0}];\n".format(qb_loc)
+
+            return f"OPENQASM 2.0;\ninclude \"qelib1.inc\";\nqreg q[{qubits}];\ncreg c[{qubits}];" \
+                   f"{self._circuit}\n{measurement_gates}"
+
+    @property
+    def _metadata(self):
+        qubit_mapping = self.main_engine.mapper.current_mapping
+        num_qubits = len(qubit_mapping.keys())
+        meas_map = [qubit_mapping[qubit_id] for qubit_id in self._measured_ids]
+
+        return {
+            "num_qubits": num_qubits,
+            "meas_map": meas_map
+        }
+
+    def estimate_cost(self, **kwargs):
+        """Estimate cost for the circuit this object has built during engine execution."""
+        return self._target.estimate_cost(
+            circuit=self._input_data,  # noqa
+            num_shots=self._num_runs,  # noqa
+            **kwargs
+        )  # noqa
+
+    def _run(self):  # pylint: disable=too-many-locals
+        """Run the circuit this object has built during engine execution."""
+        # Nothing to do with an empty circuit.
+        if not self._circuit:
+            return
+
+        if self._retrieve_execution is None:
+            res = send(
+                input_data=self._input_data,
+                metadata=self._metadata,
+                num_shots=self._num_runs,
+                target=self._target,
+                num_retries=self._num_retries,
+                interval=self._interval,
+                verbose=self._verbose
+            )
+
+            if res is None:
+                raise RuntimeError('Failed to submit job to the Azure Quantum!')
+        else:
+            res = retrieve(
+                job_id=self._retrieve_execution,
+                target=self._target,
+                num_retries=self._num_retries,
+                interval=self._interval,
+                verbose=self._verbose
+            )
+
+            if res is None:
+                raise RuntimeError(
+                    "Failed to retrieve job from Azure Quantum with job id: '{}'!".format(self._retrieve_execution)
+                )
+
+        if self._provider_id == IONQ_PROVIDER_ID:
+            self._probabilities = {
+                rearrange_result(int(k), len(self._measured_ids)): v for k, v in res["histogram"].items()
+            }
+        elif self._provider_id == QUANTINUUM_PROVIDER_ID:
+            histogram = Counter(res["c"])
+            self._probabilities = {
+                k: v / self._num_runs for k, v in histogram.items()
+            }
+
```

Comment:
Add an `else` branch here and raise an error if it is taken (ie. future-proofing the code in case we add new providers in the future).

Reply:
Updated.